src/query/processor.rs

```diff
@@ -122,14 +160,142 @@ impl<T: Transport + Clone> Processor<T> {
         let network = Network::new(self.transport.clone(), req.query_id, req.roles.clone());
         let gateway = Gateway::new(my_role, network, GatewayConfig::default()).await;
 
-        handle.set_state(QueryState::AwaitingInputs(gateway))?;
+        handle.set_state(QueryState::AwaitingInputs(req.config, gateway))?;
 
         Ok(())
     }
 
+    /// Receive inputs for the specified query. That triggers query processing
+    pub fn receive_inputs(&self, input: QueryInput) -> Result<(), QueryInputError> {
+        let mut queries = self.queries.inner.lock().unwrap();
+        match queries.entry(input.query_id) {
+            Entry::Occupied(entry) => {
+                let state = entry.remove();
+                match state {
+                    QueryState::AwaitingInputs(config, gateway) => {
+                        queries.insert(
+                            input.query_id,
+                            QueryState::Running(executor::start_query(
+                                config,
+                                gateway,
+                                input.input_stream,
+                            )),
+                        );
+                        Ok(())
+                    }
+                    _ => Err(QueryInputError::StateError {
+                        source: StateError::InvalidState {
+                            from: QueryStatus::from(&state),
+                            to: QueryStatus::Running,
+                        },
+                    }),
+                }
+            }
+            Entry::Vacant(_) => Err(QueryInputError::NoSuchQuery(input.query_id)),
+        }
+    }
+
     pub fn status(&self, query_id: QueryId) -> Option<QueryStatus> {
         self.queries.handle(query_id).status()
     }
+
+    /// Handle the next command from the input stream.
+    pub async fn handle_next(&mut self) {
+        if let Some(command) = self.command_stream.next().await {
+            match command.payload {
+                TransportCommand::Query(query_command) => match query_command {
+                    QueryCommand::Create(req, resp) => {
+                        let result = self.new_query(req).await.unwrap();
+                        resp.send(result).unwrap();
+                    }
+                    QueryCommand::Prepare(req) => {
+                        self.prepare(req).await.unwrap();
+                    }
+                    QueryCommand::Input(query_input) => {
+                        self.receive_inputs(query_input).unwrap();
+                    }
+                },
```

Comment:
i might prefer a flatter match statement here:
```suggestion
                TransportCommand::Query(QueryCommand::Create(req, resp)) => {
                    let result = self.new_query(req).await.unwrap();
                    resp.send(result).unwrap();
                },
                TransportCommand::Query(QueryCommand::Prepare(req)) => {
                    self.prepare(req).await.unwrap();
                },
                TransportCommand::Query(QueryCommand::Input(query_input)) => {
                    self.receive_inputs(query_input).unwrap();
                },
```

Reply:
I don't mind it either